{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lab 12 - RNA Sequencing through Expectation Maximization\n",
    "\n",
    "#### Authors:\n",
    "\n",
    "v1.0 (2014 Fall) Rishi Sharma \\*, Sahaana Suri \\*,  Paul Rigge \\*, Kangwook Lee \\*, Kannan Ramchandran \\* <br />\n",
    "v1.1 (2015 Fall) Kabir Chandrasekher \\*\\*, Max Kanwal \\*\\*, Kangwook Lee \\*, Kannan Ramchandran \\* <br />\n",
    "v1.2 (2016 Spring) Tony Duan \\*\\*\\*, Ashvin Nair \\*\\*\\*, David Marn \\*\\*\\*, Kabir Chandrasekher \\*\\*, Kangwook Lee \\*, Kannan Ramchandran \\*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of Contents\n",
    "\n",
    "\n",
    "- [Introduction](#Introduction)\n",
    "- [MLE for a simple model](#Question-1----Simple-Model)\n",
    "- [MLE for a harder model](#Question-2----Harder-Model)\n",
    "- [EM algorithm for the harder model](#Question-3----EM-Algorithm)\n",
    "- [References](#References)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "The problem of [RNA sequencing](http://en.wikipedia.org/wiki/RNA-Seq) is to figure out how much and what type of RNA is present in a genome at a given moment in time. It has many applications including genome annotation, comprehensive identification of fusions in cancer, discovery of novel isoforms of genes, and genome sequence assembly [[1][2]](#References).\n",
    "\n",
    "For our purposes, we'll formulate the problem as follows: given a set of short reads that are sampled from a set of larger genes, how can we find the relative abundance of each gene? That is, given just the short reads, how do we know how frequently each original gene occurs? This process is depicted in Figure 1. (Aside: in the actual paper, these \"genes\" are actually \"transcripts,\" but that's not relevant for us.)\n",
    "\n",
    "<img src=\"http://i.imgur.com/61e7d16.jpg\" title=\"source: imgur.com\" />\n",
    "\n",
    "#### <center>Figure 1: We want to use the reads to guess the underlying proportion.</center>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It turns out we can use some methods we learned in class (MLE & EM) to solve this problem -- let's try it out.\n",
    "\n",
    "We assume that all genes are of the same length, $\\ell_t$, and all reads are of the same length, $\\ell_x$. Then, we assume the following Bayesian generative model:\n",
    "\n",
    "1. A read comes from a randomly chosen gene $t_i$\n",
    "2. A read's starting point is randomly chosen among all possible $\\ell_t$ starting positions in that gene\n",
    "\n",
    "Given a set of reads $X = \\{X_1 \\ldots X_n\\}$, we want to figure out what distribution $\\rho = \\{ \\rho_1 \\ldots \\rho_m \\}$ over the genes was most likely to give us the reads $X$. To do this, we'll need to maximize the following likelihood function:\n",
    "\n",
    "$$ L(\\rho) = \\prod_{i=1}^{N}{  P(X_i|\\rho)  } $$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 1 (Optional) -- Simple Model\n",
    "\n",
    "To make life easier, we're first going to assume no read is ambiguous. That is -- given a read, we can immediately tell which gene it came from. (In practice, this means we would have to know the chart mapping each color to a gene, as in Figure 1.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1a. Suppose you are given a read $X_i$ -- what is $P(X_i|\\rho)$?\n",
    "\n",
    "Your solution should take both the gene lengths and the read lengths into consideration. Denote the probability of seeing a specific gene as $\\rho_{t_i}$.\n",
    "\n",
    "Hint: how many possible starting positions exist for $X_i$?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "= P("
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1b. Assume that you have two genes. Find the MLE of $\\rho$ if you find $x$ reads compatible with gene $1$, and $n-x$ reads compatible with gene $2$. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Your answer here*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1c. Assume you have $M$ genes. Out of $n$ reads, $x_i$ reads are compatible with gene $i$, where $\\sum_{i=1}^{M}{x_i} = n$. Find the maximum likelihood estimator of $\\rho$. \n",
    "\n",
    "Hint: you might just be able to make an \"educated guess\" from your answer above. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Your answer here.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 2 -- Harder Model\n",
    "\n",
    "Life gets harder when you allow for ambiguous reads. Going back to the figure above, we can imagine a case where we don't know which color a read belongs to, but we have a rough idea of the possible colors it could have been. See the modified figure below.\n",
    "\n",
    "<img src=\"http://i.imgur.com/5qOUtbt.jpg\" title=\"source: imgur.com\" />\n",
    "#### <center> Figure 2 </center>\n",
    "In this portion, we'll consider a general problem where a read is aligned with possibly more than one gene. \n",
    "\n",
    "First, we define a compatibility matrix $A \\in \\{0,1\\}^{n \\times m} = \\{a_{i,j}\\}$, where $a_{i,j}$ is $1$ if read $i$ is aligned with gene $j$, $0$ otherwise. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a motivating example, let's assume we have $3$ genes and $5$ reads aligned as follows:\n",
    "\n",
    "(ref: pp16-17 Pachter 2011)\n",
    "\n",
    "<img src=\"http://i.imgur.com/a0ZtnSV.png\">\n",
    "#### <center> Figure 3: Each read is a subset of some gene, and it can be compatible with multiple genes.</center>\n",
    "\n",
    "### 2a. Find the compatibility matrix $A$ for the above figure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[[1,1,1],\n",
    " [0,1,1],\n",
    " [1,0,1],\n",
    " [1,0,0],\n",
    " [1,1,0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2b. Given an arbitrary compatibility matrix, write an expression to find the likelihood function for $\\rho$.\n",
    "\n",
    "You'll have to tweak your solution to the last portion by carefully considering how you can represent $P(X_i|\\rho)$ given the compatibility matrix $A$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$=A[i,\\rho]*\\frac{1}{l_t - l_x}$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 3 -- EM Algorithm\n",
    "\n",
    "In general, it is not easy to find the exact maximum likelihood estimator of $\\rho$ if ambiguous reads exist. Instead, we can rely on iterative methods that we hope will converge to the true value. One way to go about this is via expectation maximization (EM), as you have seen in class. \n",
    "\n",
    "To recap, [here](http://ai.stanford.edu/~chuongdo/papers/em_tutorial.pdf) is a short tutorial that does a wonderful job of explaining all that you need to know about EM [[3]](#References)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For our purposes, here is an EM algorithm that can be used. You'll be implementing it shortly, so read it carefully and understand why it works. \n",
    "\n",
    "1. Initialize $\\rho = (\\frac{1}{M}, \\frac{1}{M}, \\ldots, \\frac{1}{M})$, all genes are equally probable.\n",
    "2. Find the compatibility matrix $A$.\n",
    "3. Repeat the following until $\\rho$ converges:<br/>\n",
    "    a. Each reading $i$ corresponds to the $i$th row of $A$. Call this $\\mathcal{I}_i$.<br/>\n",
    "    b. Put the values of $\\rho$ in $\\mathcal{I}_i$ where $\\mathcal{I}_i$ is not zero. Then normalize the vector and replace the $i$th row of $A$ with the normalized vector.<br/>\n",
    "    c. Replace $\\rho$ such that for each gene $j$, $\\rho_j = \\frac{1}{N}\\sum_{i=1}^{N}{A_{i,j}}$\n",
    "\n",
    "The following figure is a visual representation of the algorithm (Ref: p17 Pachter 2011) [[1]](#References)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"http://imgur.com/S0SKEzP.png\" title=\"source: imgur.com\" />\n",
    "#### <center> Figure 4: Visualization of the EM algorithm we will use. </center>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a lot to digest at once, so let's only look at the first step to begin with.\n",
    "\n",
    "There are three possible genes to analyze (red, green, blue). There are five reads (a,b,c,d,e) that align with different genes. Initially, you assume a uniform prior.\n",
    "\n",
    "1. E step. Reads are proportionately assigned to each of the genes they could have come from, according to the current distribution $\\rho$.\n",
    "2. M step. The distribution $\\rho$ is recalculated proportionally from the assigned read counts from the E step.\n",
    "\n",
    "For example, the abundance of red after the first M step is estimated by\n",
    "\n",
    "$$\\rho_1 = 0.47 = \\frac{0.33 + 0.5 + 1 + 0.5}{5}$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3a. Explain the above EM algorithm in your own words. Will it always find the MLE for $\\rho$? "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We take a guess at the initial distribution of $\\rho$, and then, using the data that we have, iteratively refine this guess until we converge at a stable value of $\\rho$. This will *not* always find the MLE for $\\rho$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3b. Implement the above EM algorithm. Run it with the given set of reads & genes. What is your estimated $\\rho$?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's begin by with a toy example. We'll start by making the relevant imports..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we'll define our genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "transcripts = ['ATCTCGACGCACTGC', 'GAGTTCGAACTCTTC', 'AGAGTTCCAGTGTCA', 'AAAGCTCACTGCGGA', 'AGCGATATCAGAGTD']\n",
    "M = len(transcripts) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll randomly generate the true distribution $\\rho$. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Container object of 5 artists>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAEACAYAAAC9Gb03AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEphJREFUeJzt3X+s3fdd3/Hny/WCcAKVgMiXOSTZkhJ+iAhaZiylsFOs\ntbdFqiv4Yw4TlYqo/EfdVjAhR0got/9siiZ1DGWweTPdWmV4osxtJApyNHQmBdTmpiSBwr3YbRVj\nJ67XVKFrmiFc570/zjfeyc21z/fenOOT8/HzIV35++PzOd/359x7X+d7Psff+01VIUlq1455FyBJ\nmi2DXpIaZ9BLUuMMeklqnEEvSY0z6CWpcb2CPslykvUkp5IcuUq7f5LkYpKf3WpfSdJsZNL/o0+y\nAzgF7AeeBVaBg1W1vkm7R4D/C/xOVf2Pvn0lSbPT54x+L3C6qs5U1UXgOHBgk3YfBD4J/O9t9JUk\nzUifoN8DnB1bP9dtuyzJPwTeU1W/DWQrfSVJszWtD2N/A3D+XZJeh3b2aPMMcOvY+i3dtnE/DhxP\nEuB7gHcm+VbPvgAk8Y/uSNIWVVX6NLrqF/AG4IvAbcANwJPAD16l/ceAn91q31Epbbr//vvnXcJM\nOb7F5vgWV5ebE3N84hl9VV1Kchg4yWiq51hVrSU51B3k6MYuk/pOfPWRJE1Nn6kbquqPgLs2bPuP\nV2j7i5P6SpKuHa+MvQYGg8G8S5gpx7fYHF/7Jl4wda0kqddLLZK0CJL0+jDWM3pJapxBL0mNM+gl\nqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNc6gl6TGGfSS1DiDXpIa\nZ9BLUuMMeklqXK+gT7KcZD3JqSRHNtn/7iRPJXkiyeNJfnps39Nj+x6bZvGSpMkm3kowyQ7gFLAf\neBZYBQ5W1fpYm11V9WK3/CPAiaq6s1v/MvCWqnp+wnG8laAkbcE0byW4FzhdVWeq6iJwHDgw3uDl\nkO/cBDw3XkvP48zd0tLtJFmor6Wl2+f9tEl6ndvZo80e4OzY+jlG4f8KSd4D/GtgCXjH2K4CHkly\nCThaVf9p++XO1oULZxiVuzguXJj4Yi7pOtcn6Hupqk8Bn0ryVuATwF3drnuq6nySmxkF/lpVPTqt\n40qSrq5P0D8D3Dq2fku3bVNV9WiSnUm+u6q+VlXnu+1fTXKC0buBTYN+ZWXl8vJgMGAwGPQoT5Ku\nD8PhkOFwuOV+fT6MfQPw14w+jD0PPAbcW1VrY23uqKovdctvBn6vqu5IsgvYUVUvJLkROAl8pKpO\nbnKcuX8Ym4RFm7qBMO/nTdJ89P0wduIZfVVdSnKYUUjvAI5V1VqSQ6PddRT4uSTvBf4e+Cbwz7vu\nu4ETSao71kObhbwkaXYmntFfK57Rb5dn9NL1apr/vVKStMAMeklqnEEvSY0z6CWpcQa9JDXOoJek\nxhn0ktQ4g16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWqc\nQS9JjesV9EmWk6wnOZXkyCb7353kqSRPJHk8yU/37StJmq2JNwdPsgM4BewHngVWgYNVtT7WZldV\nvdgt/whwoqru7NN37DG8Ofi2eHNw6Xo1zZuD7wVOV9WZqroIHAcOjDd4OeQ7NwHP9e0rSZqtPkG/\nBzg7tn6u2/YKSd6TZA34DPChrfSVJM3Ozmk9UFV9CvhUkp8EPgHctdXHWFlZubw8GAwYDAbTKk+S\nFt5wOGQ4HG65X585+n3ASlUtd+v3AVVVD1ylz5cYTdu8qW9f5+i3yzl66Xo1zTn6VeDOJLcluQE4\nCDy84WB3jC2/GaCqvtanryRptiZO3VTVpSSHgZOMXhiOVdVakkOj3XUU+Lkk7wX+Hvgmo0C/Yt8Z\njUWStImJUzfXilM32+XUjXS9mubUjSRpgRn0ktQ4g16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1\nzqCXpMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjTPoJalxBr0kNc6gl6TGGfSS1DiDXpIa1yvokywn\nWU9yKsmRTfb/fJKnuq9Hk9w9tu/pbvsTSR6bZvGSpMl2TmqQZAfwILAfeBZYTfLpqlofa/Zl4Keq\n6utJloGjwL5u30vAoKqen27pkqQ++pzR7wVOV9WZqroIHAcOjDeoqs9W1de71c8Ce8Z2p+dxJEkz\n0CeA9wBnx9bP8cog3+iXgD8cWy/gkSSrSd6/9RIlSa/FxKmbrUjyNuB9wFvHNt9TVeeT3Mwo8Neq\n6tHN+q+srFxeHgwGDAaDaZYnSQttOBwyHA633C9VdfUGyT5gpaqWu/X7gKqqBza0uxv4fWC5qr50\nhce6H/hGVX10k301qZZZS8LoDcgiCfN+3iTNRxKqKpPa9Zm6WQXuTHJbkhuAg8DDGw52K6OQ/4Xx\nkE+yK8lN3fKNwNuBL/QfhiTptZo4dVNVl5IcBk4yemE4VlVrSQ6NdtdR4NeB7wJ+K6PT4otVtRfY\nDZxIUt2xHqqqk7MajCTp1SZO3VwrTt1sl1M30vVqmlM3kqQFZtBLUuMMeklqnEEvSY0z6CWpcQa9\nJDXOoJekxhn0ktQ4g16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS\n4wx6SWpcr6BPspxkPcmpJEc22f/zSZ7qvh5NcnffvpKk2Zp4c/AkO4BTwH7gWWAVOFhV62Nt9gFr\nVfX1JMvASlXt69N37DG8Ofi2eHNw6Xo1zZuD7wVOV9WZqroIHAcOjDeoqs9W1de71c8Ce/r2lSTN\nVp+g3wOcHVs/x/8P8s38EvCH2+wrSZqyndN8sCRvA94HvHU7/VdWVi4vDwYDBoPBVOqSpBYMh0OG\nw+GW+/WZo9/HaM59uVu/D6iqemBDu7uB3weWq+pLW+nb7XOOfluco5euV9Oco18F7kxyW5IbgIPA\nwxsOdiujkP+Fl0O+b19J0mxNnLqpqktJDgMnGb0wHKuqtSSHRrvrKPDrwHcBv5XRafHFqtp7pb4z\nG40k6VUmTt1cK07dbJdTN9L1appTN5KkBWbQX0eWlm4nycJ8LS3dPu+nTGqCUzevrIGWp24Wb3xO\nS0lX49SNrju+Y5E25xn9K2tgsc54wTP6sdaNj0/ayDN6SRJg0EtS8wx6SWqcQS9JjTPoJalxBr0k\nNc6gl6TGGfSS1DiDXpIaZ9BLUuMMeklqnEEvSY0z6CWpcQa9JDWuV9AnWU6ynuRUkiOb7L8ryZ8m\n+bskv7Jh39NJnkryRJLHplW4JKmfnZMaJNkBPAjsB54FVpN8uqrWx5p9Dfgg8J5NHuIlYFBVz0+h\nXknSFvU5o98LnK6qM1V1ETgOHBhvUFXPVdXngW9t0j89jyNJmoE+AbwHODu2fq7b1lcBjyRZTfL+\nrRQnSXrtJk7dTME9VXU+yc2MAn+tqh7drOHKysrl5cFgwGAwuAblSdJiGA6HDIfDLfebeM/YJPuA\nlapa7tbvA6qqHtik7f3AN6rqo1d4rCvu956x2+U9Yy+3bnx80kbTvGfsKnBnktuS3AAcBB6+2rHH\nitiV5KZu+Ubg7cAXehxTkjQlE6duqupSksPASUYvDMeqai3JodHuOppkN/A48B3AS0k+DPwQcDNw\nIkl1x3qoqk7OajCSpFebOHVzrTh1s11O3Vxu3fj4Wra0dDsXLpyZdxlbsnv3bXzlK0/PtYa+UzcG\n/StrYLGCAgz6sdaNj69li/e9g9fD92+ac/SSpAVm0EtS4wx6SWqcQS9JjTPoJalxBr0kNc6gl6TG\nGfSS1DiDXloAS0u3k2ShvpaWbp/306aOV8a+sgZavjpv8cbnlbGXWy7c2MDxXYMKvDJWkgQGvSQ1\nz6CXpMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjesV9EmWk6wnOZXkyCb770ryp0n+LsmvbKWvJGm2\nJgZ9kh3Ag8A7gB8G7k3yAxuafQ34IPBvttFXkjRDfc7o9wKnq+pMVV0EjgMHxhtU1XNV9XngW1vt\nK0marT5Bvwc4O7Z+rtvWx2vpK0magp3zLmDcysrK5eXBYMBgMJhbLZL0ejMcDhkOh1vuN/GvVybZ\nB6xU1XK3fh9QVfXAJm3vB75RVR/dRl//euW2tPwXAv3rlZdbLtzYwPFdgwqm+NcrV4E7k9yW5Abg\nIPDw1Y79GvpKkqZs4tRNVV1Kchg4yeiF4VhVrSU5NNpdR5PsBh4HvgN4KcmHgR+qqhc26zuz0UiS\nXsUbj7yyBlp++7h443Pq5nLLhRsbOL5rUIE3HpEkgUEvSc0z6CWpcQa9JDXOoJekxhn0ktQ4g16S\nGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjesV9EmW\nk6wnOZXkyBXa/GaS00meTPJjY9ufTvJUkieSPDatwiVJ/Uy8OXiSHcCDwH7gWWA1yaeran2szTuB\nO6rqTUl+AvhtYF+3+yVgUFXPT716SdJEfc7o9wKnq+pMVV0EjgMHNrQ5AHwcoKo+B7wxye5uX3oe\nR5I0A30CeA9wdmz9XLftam2eGWtTwCNJVpO8f7uFSpK2Z+LUzRTcU1Xnk9zMKPDXqurRa3BcSRL9\ngv4Z4Nax9Vu6bRvbfN9mbarqfPfvV5OcYDQVtGnQr6ysXF4eDAYMBoMe5UnS9WE4HDIcDrfcL1V1\n9QbJG4C/ZvRh7HngMeDeqloba/Mu4ANV9TNJ9gG/UVX7kuwCdlTVC0luBE4CH6mqk5scpybVMmtJ\nGM00LZLQ93lbvPH1Hxu0Pb7FGxs4vmtQQUJVZVK7iWf0VXUpyWFGIb0DOFZVa0kOjXbX0ar6TJJ3\nJfki8E3gfV333cCJJNUd66HNQl6SNDsTz+ivFc/ot6vlsybP6C+3XLixgeO7BhX0PKP3vz1KUuMM\neklqnEEvSY0z6CWpcQa9JDXOoJekxhn0ktQ4g16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCX\npMYZ9JLUOINekhpn0EtS4wx6SWqcQS9JjesV9EmWk6wnOZXkyBXa/GaS00meTPKjW+krSZqdiUGf\nZAfwIPAO4IeBe5P8wIY27wTuqKo3AYeA/9C37/VhOO8CZmw47wJmbDjvAmZsOO8CZmw47wLmrs8Z\n/V7gdFWdqaqLwHHgwIY2B4CPA1TV54A3Jtnds+91YDjvAmZsOO8CZmw47wJmbDjvAmZsOO8C5q5P\n0O8Bzo6tn+u29WnTp68kaYZm9WFsZvS4kqQt2tmjzTPArWPrt3TbNrb5vk3a3NCj72XJ6+H1YVY1\nfGRGj7vV522xxrf1n4mWxzfL3w/Htx2vj8yarE/QrwJ3JrkNOA8cBO7d0OZh4APAf0+yD/jbqrqQ\n5LkefQGoqsV4xiRpwUwM+qq6lOQwcJLRVM+xqlpLcmi0u45W1WeSvCvJF4FvAu+7Wt+ZjUaS9Cqp\nqnnXIEmaoblfGdvyBVVJjiW5kOTP513LLCS5JckfJ/nLJH+R5EPzrmmaknxbks8leaIb47+ad03T\nlmRHkj9L8vC8a5m2JE8near7/j0273qmLckbk/xekrXu5/Mnrth2nmf03QVVp4D9wLOMPg84WFXr\ncytqipK8FXgB+HhV3T3veqYtyRKwVFVPJrkJ+DxwoJXvH0CSXVX1YpI3AH8C/Muq+pN51zUtSX4Z\neAvwnVX17nnXM01Jvgy8paqen3cts5DkvwD/q6o+lmQnsKuq/s9mbed9Rt/0BVVV9SjQ5A8ZQFV9\npaqe7JZfANZo7DqJqnqxW/w2Rr8vzXw/k9wCvAv4z/OuZUbC/DNuJpJ8J/CTVfUxgKr61pVCHub/\nJHhBVSOS3A78KPC5+VYyXd3UxhPAV4BhVf3VvGuaon8L/CrQ6gd1BTySZDXJ++ddzJT9I+C5JB/r\npt6OJvn2KzWed9CrAd20zSeBD3dn9s2oqpeq6scYXQPyU0n+6bxrmoYkPwNc6N6RhTYvcrynqt7M\n6F3LB7qp1FbsBN4M/PtujC8C912p8byDvs/FWHod6+YGPwl8oqo+Pe96ZqV7W/wHwI/Pu5YpuQd4\ndzeP/bvA25J8fM41TVVVne/+/SpwgtFUcSvOAWer6vFu/ZOMgn9T8w76yxdjJbmB0QVVrX363+rZ\n0st+B/irqvp38y5k2pJ8T5I3dsvfDvwz4Mn5VjUdVfVrVXVrVf1jRr93f1xV7513XdOSZFf3TpMk\nNwJvB74w36qmp6ouAGeTfH+3aT9wxWnFPlfGzkzrF1Ql+W/AAPjuJH8D3P/yhyctSHIP8C+Av+jm\nsQv4tar6o/lWNjXfC/zXjK5z38HoXcv/nHNN6mc3cCJJMcq5h6rq5JxrmrYPAQ8l+QfAl+kuVN2M\nF0xJUuPmPXUjSZoxg16SGmfQS1LjDHpJapxBL0mNM+glqXEGvSQ1zqCXpMb9P+86icN2Yu9RAAAA\nAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10d59b4e0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rho = np.random.rand(M)\n",
    "rho /= sum(rho)\n",
    "plt.bar(np.arange(M) + 0.6, rho, width=0.8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll then pick 1000 random reads of length 5 each."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 10 reads... ['GAGTT', 'CGCAC', 'CTGCG', 'TTCCA', 'GAGTT', 'CGCAC', 'GAGTT', 'CGAAC', 'CCAGT', 'ACTCT']\n"
     ]
    }
   ],
   "source": [
    "def random_read( s, rho, L ):\n",
    "    chosen_seq = np.random.choice(s, p=rho)\n",
    "    start_idx = np.random.randint(0, len(chosen_seq) - L)\n",
    "    end_idx = start_idx + L\n",
    "    return chosen_seq[ start_idx:end_idx ]\n",
    "        \n",
    "N = 1000 # Number of reads\n",
    "L = 5\n",
    "\n",
    "reads = []\n",
    "for i in range(N):\n",
    "    reads.append( random_read(transcripts, rho, L) )\n",
    "    \n",
    "print('First 10 reads...', reads[0:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the code below, run the EM algorithm for 100 iterations and report your estimated distribution of $\\rho$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "N_iter = 100 #number of E/M iterations\n",
    "\n",
    "def find_all_alignments(s, read):\n",
    "    tmp = []\n",
    "    for j in range(len(s)):\n",
    "        if read in s[j]:\n",
    "            tmp.append(j)\n",
    "    return tmp\n",
    "\n",
    "# A: Compatibility Matrix. \n",
    "# Note: You can choose to represent this matrix in whatever form is easiest for your calculations\n",
    "# The form we are pushing you towards here is just how we chose to implement the algorithm\n",
    "A = []\n",
    "for each_read in reads:\n",
    "    A.append( find_all_alignments(transcripts, each_read) )\n",
    "\n",
    "hidden_state_prior = np.zeros([N, M])\n",
    "rho_est = np.array([1/5, 1/5, 1/5, 1/5, 1/5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for i in range(N_iter):\n",
    "    # 1. E step\n",
    "    for i in range(N):\n",
    "        validGenes = A[i]\n",
    "        for j in range(M):\n",
    "            if j in validGenes:\n",
    "                hidden_state_prior[i,j] = rho_est[j]*(1/10)\n",
    "            else:\n",
    "                hidden_state_prior[i,j] = 0    \n",
    "        hidden_state_prior[i] /= sum(hidden_state_prior[i])\n",
    "            \n",
    "\n",
    "        \n",
    "                \n",
    "    # 2. M step\n",
    "    rho_est = hidden_state_prior.sum(axis=0)\n",
    "    rho_est /= sum(rho_est)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Your estimated distribution should look similar to the real distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(real) rho [ 0.3553295   0.18893396  0.18659584  0.12549704  0.14364365]\n",
      "(est.) rho [ 0.36607565  0.19969755  0.17446452  0.10892435  0.15083793]\n"
     ]
    }
   ],
   "source": [
    "print('(real) rho', rho)\n",
    "print('(est.) rho', rho_est)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAEACAYAAAC9Gb03AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAGbtJREFUeJzt3X2UVPWd5/H3pyF4RAVBpJs00I3ikkwi68MscuJTMSSK\nzkHMTMz6cDSisuQBDZtkg3kwNolnjGwyOxp2ZiWDbpLDLnN8CphExYmn4sGNoXWQEKYREOlAg42A\nCGiIBL77R5Wdoino21CXoi+f1zl9uA+/373f28Cnbv3q3luKCMzMLLtqql2AmZmly0FvZpZxDnoz\ns4xz0JuZZZyD3sws4xz0ZmYZlyjoJU2QtFLSKkkzDtHuP0naI+lvutvXzMzSoa6uo5dUA6wCxgMb\ngWbg2ohYWabds8AfgIci4vGkfc3MLD1JzujHAKsjojUi9gDzgUll2t0OPApsPoy+ZmaWkiRBXw+s\nL5nfUFzWQdIHgasj4p8AdaevmZmlq1Ifxv4D4PF3M7NjUO8EbdqA4SXzQ4vLSv0lMF+SgEHAFZL+\nlLAvAJL80B0zs26KCCVpdMgfoBewBmgA+gCvAB8+RPuHgb/pbt9CKdl09913V7uEVPn4ejYfX89V\nzM0uc7zLM/qI2CtpGrCIwlDP3IhokTS1uJM5nbt01bfLVx8zM6uYJEM3RMTTwKhOyx48SNtbuupr\nZmZHj++MPQpyuVy1S0iVj69n8/FlX5c3TB0tkuJYqcXMrCeQlOjD2ERDN2aWbY2NjbS2tla7DDuI\nhoYG1q1bd9j9fUZvZu+fGVa7DDuIg/39JD2j9xi9mVnGOejNzDLOQW9mlnEOejM7bo0bN46HHnqo\n2mWkzkFvZmXV1TUiKbWfurrGxLU0NjbSt29f+vXrx5AhQ7jxxhvZuXNnegefMQ56Myurvb2VwhNN\n0vkpbD8ZSfz85z9nx44dLFu2jOXLl3PPPfdU4jCPCw56M+sR3r+8cPDgwVx++eWsWLECgPfee4+v\nfOUrNDQ0MGTIED7/+c/zxz/+EYDt27czceJEBg8ezGmnncbEiRNpayv7AN1Mc9CbWY+yYcMGnnrq\nKS644AIAZsyYwZo1a/jtb3/LmjVraGtr49vf/jYA+/bt45ZbbmH9+vX8/ve/p2/fvkybNq2a5VeF\nb5gys7I35BS+XiLN/5PJb9IaMWIEW7duBWDXrl1MmjSJxx57jJqaGk4++WSWL1/OiBEjAPj1r3/N\nDTfcwNq1aw/YziuvvML48eM7tjVu3DhuvPFGbrnllgPaHkuO9IYpPwLBzHqEBQsWMG7cOJ5//nmu\nuuoqXn75ZRobG3n33Xc5//zzO9rt27evIxT/8Ic/MH36dJ555hm2b99ORLBr1y4iovhCdnzw0I2Z\n9Qjvh/cll1zCtGnTmDFjBoMGDaJv376sWLGCbdu2sW3bNrZv387bb78NwPe//31Wr15Nc3Mz27dv\n5/nnn99vW8cLB/1RUDe0Lr1L1IbWVfvwzI666dOns2TJEpYsWcKUKVOYPn06b775JgBtbW0sWrQI\ngJ07d3LiiSfSr18/tm3bRlNTUxWrrh4P3RwF7W3t0JTStpva09mwHfdqaxtob09veKO2tiFx287D\nLIMGDeIzn/kM9913H/Pnz2fmzJmMHTuWrVu3Ul9fz+c+9zkuu+wypk+fzvXXX8+gQYOor6/ny1/+\nMgsXLjzodrPKH8YeBZJSC3qajr+3oVZ5fnrlsc1PrzQzs0NKFPSSJkhaKWmVpBll1l8laZmkpZJe\nkvRXJevWlaxbUsnizcysa12O0UuqAWYD44GNQLOkBRGxsqTZv0bEwmL7s4EngJHFdfuAXES8VdHK\nzcwskSRn9GOA1RHRGhF7gPnApNIGEfFuyezJwJaSeSXcT9Wl9RAnM7NqShLA9cD6kvkNxWX7kXS1\npBbgF8AdJasCeFZSs6QpR1Js2tJ7iJOZWfVU7PLKiPgp8FNJFwE/AUYVV10YEZsknU4h8FsiYnGl\n9mtmZoeWJOjbgOEl80OLy8qKiMWSeks6LSK2RsSm4vI3JT1BYSiobNCX3syQy+XI5XIJyjMzOz7k\n83ny+Xy3+3V5Hb2kXsCrFD6M3QQsAa6LiJaSNmdGxGvF6fOARyLiTEl9gZqI2CXpJGARMDMiFpXZ\nT9Wvo0/vIU6+jt6Obb6O/tiW+nX0EbEXmEYhpFcA8yOiRdJUSf+l2OxvJf1O0r8B9wP/ubi8Flgs\naSnwIvBkuZA3M0vb4sWL+fCHP1ztMsr61a9+xbBhw1LbfqIx+oh4mj+Pub+/7MGS6VnArDL9XgfO\nOcIazawK6obWFR7fkZLa+lre2PBGoraNjY1s3ryZ3r17dzx58uabb+aBBx44aJ+amhrWrFnDGWec\nAcBFF11ES0vLQdsficmTJzNs2LCO5+AfjjSv0POzbsysrDSf0QTde07T+18lOG7cuG71sYIecX27\nmVm5MerXXnuNXC7HqaeeyuDBg7nuuusAuPTSS4kIRo8eTb9+/XjkkUcOGB4ZMWIE3/ve9zra3Hbb\nbWzevJkrr7yS/v37c9lll3U87hjg05/+NEOGDGHAgAHkcrmOdwc//OEPmTdvHrNmzaJfv35MmlS4\nzWjTpk186lOfYvDgwZx55pn84Ac/6NjW7t27ufnmmxk4cCAf/ehHaW5uTuV39j4HvZn1WHfddReX\nX34527dvZ8OGDdx+++1AYcwbYPny5ezYsYNrrrkGOPAs//HHH+e5557j1Vdf5cknn+SKK67gu9/9\nLm+++SZ79+7db2joyiuv5LXXXmPz5s2cd955XH/99QBMmTKFG264ga9+9avs2LGDBQsWEBFMnDiR\nc889l02bNvHLX/6S+++/n2effRYoXGH4+uuv8/rrr/PMM8/wox/9KNXfk4PezHqEq6++moEDBzJg\nwAAGDhzI3Llz6dOnD62trbS1tdGnTx8+9rGP7denqyuJbr/9dgYNGsSQIUO4+OKLGTt2LKNHj6ZP\nnz588pOfZOnSpR1tb775Zvr27csHPvABvvWtb7Fs2TJ27txZdrvNzc1s2bKFb3zjG/Tq1YvGxkZu\nu+025s+fD8AjjzzCN7/5Tfr37099fT133HFH2e1UioPezHqEBQsWsG3bNt566y22bdvGrbfeyqxZ\ns9i3bx9jxozh7LPP5uGHH+7WNmtrazumTzzxxAPmd+3aBRS+nvDOO+9k5MiRnHrqqYwYMQJJbNmy\n5YBtAh0vPgMHDux4cbr33nvZvHkzABs3bmTo0KEd7Rsakj+b/3D4w1gz6xHKnZ0PHjyYOXPmAPDC\nCy/w8Y9/nEsvvbTjSptKmTdvHk8++STPPfccw4cP5+2332bAgAEdNXUeEho2bBhnnHEGr776atnt\nffCDH2T9+vUdl3u2trZWtN7OfEZvZj3Wo48+Sltb4Ub9U089lZqaGmpqCrFWV1fH2rVrK7KfXbt2\nccIJJzBgwADeeecdvva1r+0X7rW1tfvta8yYMZxyyinMmjWL3bt3s3fvXlasWMFLL70EwDXXXMO9\n997b8dnC7NmzK1LnwfiM3szKqq2vTfWrKmvra7tuVGLixIn06tWr4zr6T3ziE5x11llMnz6dHTt2\nUFtbywMPPEBjYyNQ+MDzpptuYvfu3cyZM4fTTz99v+11Pgs/1OWYN910E8888wz19fWcdtppfOc7\n3+HBBztuJeLWW2/lmmuuYeDAgeRyOR5//HF+9rOf8aUvfYkRI0bw3nvvMWrUKO655x4A7r77bj77\n2c8yYsQI6uvrmTx5Mvfff3+3fh/d4a8S3L8G/AgEOx75EQjHNn+VoJmZHZKD3sws4xz0ZmYZ56A3\nM8s4B72ZWcY56M3MMs7X0ZsZDQ0NfqzvMexIH5HgoDcz1q1bV+0SLEUeujEzyzgHvZlZxiUKekkT\nJK2UtErSjDLrr5K0TNJSSS9J+qukfc3MLF1djtFLqgFmA+OBjUCzpAURsbKk2b9GxMJi+7OBJ4CR\nCfuamVmKkpzRjwFWR0RrROwB5gOTShtExLslsycDW5L2NTOzdCUJ+npgfcn8huKy/Ui6WlIL8Avg\nju70NTOz9FTs8sqI+CnwU0kXAz8BRnV3G01NTR3TuVyOXC5XqfLMzHq8fD5PPp/vdr8un0cvaSzQ\nFBETivN3AhER9x2iz2sUhm3OStrXz6M/TE1+Hr3Z8aqSz6NvpvDBaoOkPsC1wMJOOzuzZPo8gIjY\nmqSvmZmlq8uhm4jYK2kasIjCC8PciGiRNLWwOuYAfyvpJuA94B0KgX7Qvikdi5mZleGvEty/Bjx0\nY2Y9hb9K0MzMAAe9mVnmOejNzDLOQW9mlnEOejOzjHPQm5llnIPezCzjHPRmZhnnoDczyzgHvZlZ\nxjnozcwyzkFvZpZxDnozs4xz0JuZZZyD3sws4xz0ZmYZ56A3M8s4B72ZWcY56M3MMi5R0EuaIGml\npFWSZpRZf72kZcWfxZJGl6xbV1y+VNKSShZvZmZd691VA0k1wGxgPLARaJa0ICJWljRbC1wSEW9L\nmgDMAcYW1+0DchHxVmVLNzOzJJKc0Y8BVkdEa0TsAeYDk0obRMSLEfF2cfZFoL5ktRLux8zMUpAk\ngOuB9SXzG9g/yDu7DXiqZD6AZyU1S5rS/RLNzOxIdDl00x2SxgGTgYtKFl8YEZsknU4h8FsiYnG5\n/k1NTR3TuVyOXC5XyfLMzHq0fD5PPp/vdj9FxKEbSGOBpoiYUJy/E4iIuK9Tu9HAY8CEiHjtINu6\nG9gZEX9fZl10VUvaJFF4A1LxLUNTCpsFaIJq/97MrDokERHqql2SoZtmYKSkBkl9gGuBhZ12NpxC\nyN9YGvKS+ko6uTh9EnAZ8Lvkh2FmZkeqy6GbiNgraRqwiMILw9yIaJE0tbA65gB3AQOBf1ThtHhP\nRIwBaoEnJEVxX/MiYlFaB2NmZgfqcujmaPHQzWFq8tCN2fGqkkM3ZmbWgznozcwyzkFvZpZxDnoz\ns4xz0JuZZZyD3sws4xz0ZmYZ56A3M8s4B72ZWcY56M3MMs5Bb2aWcQ56M7OMc9CbmWWcg97MLOMc\n9GZmGeegNzPLOAe9mVnGOejNzDLOQW9mlnGJgl7SBEkrJa2SNKPM+uslLSv+LJY0OmlfMzNLV5dB\nL6kGmA1cDnwEuE7Shzo1WwtcEhH/EbgHmNONvmZmlqIkZ/RjgNUR0RoRe4D5wKTSBhHxYkS8XZx9\nEahP2tfMzNKVJOjrgfUl8xv4c5CXcxvw1GH2NTOzCutdyY1JGgdMBi46nP5NTU0d07lcjlwuV5G6\nLF11Q+tob2uv+HZr62t5Y8MbFd+uWU+Vz+fJ5/Pd7qeIOHQDaSzQFBETivN3AhER93VqNxp4DJgQ\nEa91p29xXXRVS9okAWnUIGhKYbMATXBM/N6aUthwU/WPzexYJomIUFftkgzdNAMjJTVI6gNcCyzs\ntLPhFEL+xvdDPmlfMzNLV5dDNxGxV9I0YBGFF4a5EdEiaWphdcwB7gIGAv+owmnxnogYc7C+qR2N\nmZkdINEYfUQ8DYzqtOzBkukpwJSkfc3M7OjxnbFmZhnnoD+O1NU1IqniP2Z2bKvo5ZV2bGtvbyW1\nq4rM7JjlM3rLjLTesdTVNVb70MyOiM/oLTPSesfS3u53LNaz+YzezCzjHPRmZhnnoDczyzgHvZlZ\nxjnozbrSi3Su5hlaV+0js+OEr7ox68peUnk6Z3tT5R/tbFaOz+jNzDLOQW9mlnEOejOzjHPQm5ll\nnIPezCzjHPRmZhnnoDczyzgHvZlZxiUKekkTJK2UtErSjDLrR0n6f5J2S/pSp3XrJC2TtFTSkkoV\nbmZmyXR5Z6ykGmA2MB7YCDRLWhARK0uabQVuB64us4l9QC4i3qpAvWZm1k1JzujHAKsjojUi9gDz\ngUmlDSJiS0S8DPypTH8l3I+ZmaUgSQDXA+tL5jcUlyUVwLOSmiVN6U5xZmZ25I7GQ80ujIhNkk6n\nEPgtEbG4XMOmpqaO6VwuRy6XOwrlmVmW1Q2to72t8g+Qq62v5Y0Nb1R8u4eSz+fJ5/Pd7pck6NuA\n4SXzQ4vLEomITcU/35T0BIWhoC6D3sysEtrb2jPz9NHOJ8AzZ85M1C/J0E0zMFJSg6Q+wLXAwkO0\n7/gmZUl9JZ1cnD4JuAz4XaLKzMysIro8o4+IvZKmAYsovDDMjYgWSVMLq2OOpFrgJeAUYJ+kLwJ/\nAZwOPCEpivuaFxGL0joYMzM7UKIx+oh4GhjVadmDJdPtwLAyXXcB5xxJgWaWfXV1jbS3t1a7jMzy\nN0yZWdUVQj5S2rq6bpJxvr7dzCzjHPRmZhnnoDczyzgHvZlZxjnozcwyzkFvZpZxDnozs4xz0JuZ\nZZyD3qwHqKtrRFIqP3V1jdU+PEuZ74w16wHSvHO0vd13jmadz+jNzDLOQW9mlnEOejOzjHPQm5ll\nnIPezCzjHPRmZhnnoDczyzgHvdnxrhfp3Ig1tK7aR2ZFiW6YkjQB+Af+/OXg93VaPwp4GDgP+HpE\n/H3SvmZWZXuBpspvtr2pvfIbtcPS5Rm9pBpgNnA58BHgOkkf6tRsK3A78N8Po6+ZmaUoydDNGGB1\nRLRGxB5gPjCptEFEbImIl4E/dbevmZmlK0nQ1wPrS+Y3FJclcSR9zcysAo6ph5o1NTV1TOdyOXK5\nXNVqMTM71uTzefL5fLf7JQn6NmB4yfzQ4rIkutW3NOjNzGx/nU+AZ86cmahfkqGbZmCkpAZJfYBr\ngYWHaF/6zNPu9jUzswrr8ow+IvZKmgYs4s+XSLZImlpYHXMk1QIvAacA+yR9EfiLiNhVrm9qR2Nm\nZgdINEYfEU8Dozote7Bkuh0YlrSvmZkdPb4z1sws4xz0ZmYZ56A3M8s4B72ZWcY56M3MMs5Bb2aW\ncQ56M7OMc9CbmWWcg97MLOMc9GZmGeegNzPLOAe9mVnGOejNzDLOQW9mlnEOejOzjHPQm5llnIPe\nzCzjHPRmZhmXKOglTZC0UtIqSTMO0uYBSaslvSLp3JLl6yQtk7RU0pJKFW5mZsl0+Z2xkmqA2cB4\nYCPQLGlBRKwsaXMFcGZEnCXpAuCfgLHF1fuAXES8VfHqzcysS0nO6McAqyOiNSL2APOBSZ3aTAJ+\nDBARvwH6S6otrlPC/ZiZWQqSBHA9sL5kfkNx2aHatJW0CeBZSc2SphxuoWZmdni6HLqpgAsjYpOk\n0ykEfktELD4K+zUzM5IFfRswvGR+aHFZ5zbDyrWJiE3FP9+U9ASFoaCyQd/U1NQxncvlyOVyCcoz\nMzs+5PN58vl8t/slCfpmYKSkBmATcC1wXac2C4EvAP8iaSywPSLaJfUFaiJil6STgMuAmQfbUWnQ\nm5nZ/jqfAM+cedA43U+XQR8ReyVNAxZRGNOfGxEtkqYWVseciPiFpCslrQHeASYXu9cCT0iK4r7m\nRcSibhyXmZkdoURj9BHxNDCq07IHO81PK9PvdeCcIynQzMyOjC97NDPLOAe9mVnGOejNzDLOQW9m\nlnEOejOzjHPQm5llnIPezCzjHPRmZhnnoDczyzgHvZlZxjnozcwyzkFvZpZxDnozs4xz0JuZZZyD\n3sws4xz0ZmYZ56A3M8s4B72ZWcY56M3MMi5R0EuaIGmlpFWSZhykzQOSVkt6RdI53elrZmbp6TLo\nJdUAs4HLgY8A10n6UKc2VwBnRsRZwFTgfyXta2Zm6UpyRj8GWB0RrRGxB5gPTOrUZhLwY4CI+A3Q\nX1Jtwr5mZpaiJEFfD6wvmd9QXJakTZK+ZmaWorQ+jFVK2zUzs27qnaBNGzC8ZH5ocVnnNsPKtOmT\noG8H6Vh4fUiphqZ0Ngvd/b31rOPr/r+JLB9fiv8/mtLZrI/v2JAk6JuBkZIagE3AtcB1ndosBL4A\n/IukscD2iGiXtCVBXwAiomf8xszMepgugz4i9kqaBiyiMNQzNyJaJE0trI45EfELSVdKWgO8A0w+\nVN/UjsbMzA6giKh2DWZmlqKq3xmb5RuqJM2V1C7pt9WuJQ2Shkp6TtIKScsl3VHtmipJ0gmSfiNp\nafEY/67aNVWapBpJ/yZpYbVrqTRJ6yQtK/79Lal2PZUmqb+kRyS1FP99XnDQttU8oy/eULUKGA9s\npPB5wLURsbJqRVWQpIuAXcCPI2J0teupNEl1QF1EvCLpZOBlYFJW/v4AJPWNiHcl9QJeAL4cES9U\nu65KkfRfgfOBfhFxVbXrqSRJa4HzI+KtateSBkn/G/hVRDwsqTfQNyJ2lGtb7TP6TN9QFRGLgUz+\nIwOIiDci4pXi9C6ghYzdJxER7xYnT6Dw/yUzf5+ShgJXAv9c7VpSIqqfcamQ1A+4OCIeBoiIPx0s\n5KH6vwTfUJURkhqBc4DfVLeSyioObSwF3gDyEfHv1a6pgv4H8N+ArH5QF8CzkpolTal2MRU2Atgi\n6eHi0NscSScerHG1g94yoDhs8yjwxeKZfWZExL6IOJfCPSCXSLq02jVVgqS/BtqL78hENm9yvDAi\nzqPwruULxaHUrOgNnAf8z+IxvgvcebDG1Q76JDdj2TGsODb4KPCTiFhQ7XrSUnxb/HPgL6tdS4Vc\nCFxVHMf+v8A4ST+uck0VFRGbin++CTxBYag4KzYA6yPipeL8oxSCv6xqB33HzViS+lC4oSprn/5n\n9WzpfQ8B/x4R91e7kEqTNEhS/+L0icAngFeqW1VlRMTXI2J4RJxB4f/dcxFxU7XrqhRJfYvvNJF0\nEnAZ8LvqVlU5EdEOrJf0H4qLxgMHHVZMcmdsarJ+Q5Wk/wPkgNMk/R64+/0PT7JA0oXADcDy4jh2\nAF+PiKerW1nFDAF+pMJ97jUU3rX8sso1WTK1wBOSgkLOzYuIRVWuqdLuAOZJ+gCwluKNquX4hikz\ns4yr9tCNmZmlzEFvZpZxDnozs4xz0JuZZZyD3sws4xz0ZmYZ56A3M8s4B72ZWcb9f/pkcm8POJKD\nAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x107a9a6d8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.bar(np.arange(M) + 0.6, rho, color='blue', width=0.4, label='Real')\n",
    "plt.bar(np.arange(M) + 1, rho_est, color='green', width=0.4, label='Estimated')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3c. Run the same EM algorithm on our bigger dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's try running the same algorithm on much more data. Here we have a dataset of 150 transcripts of length varying from 4 to 19. We've gathered 10000 reads from exactly 5 of them -- can you recover which 5 transcripts we gathered the reads from?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "M, N = 150, 10000\n",
    "transcripts = np.load('transcripts.npy')\n",
    "reads = np.load('reads.npy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATCTCGACGCACTGC', 'GAGTTCGAACTCTTC', 'AGAGTTCCAGTGTCA', 'AAAGCTCACTGCGGA', 'AGCGATATCAGAGTD']\n",
      "['GAGTT', 'CGCAC', 'CTGCG', 'TTCCA', 'GAGTT', 'CGCAC', 'GAGTT', 'CGAAC', 'CCAGT', 'ACTCT', 'TTCGA', 'TCGAC', 'ATCTC', 'TCGAA', 'CACTG', 'CTGCG', 'CACTG', 'TCCAG', 'GAGTT', 'GAACT', 'ATCTC', 'CTCAC', 'TCGAC', 'CGACG', 'TTCGA', 'AGTTC', 'GCACT', 'CGCAC', 'TTCCA', 'TCTCG', 'CTCTT', 'GACGC', 'TCTCG', 'TCGAA', 'TATCA', 'CCAGT', 'TATCA', 'TCGAA', 'CACTG', 'CACTG', 'AGAGT', 'TCTCG', 'CTCAC', 'TCAGA', 'ATCTC', 'ATATC', 'AGTTC', 'ACGCA', 'AGTTC', 'AGTTC', 'CTCGA', 'CACTG', 'CGACG', 'CGCAC', 'CGAAC', 'TCGAC', 'AGTTC', 'TCCAG', 'CACTG', 'AAAGC', 'GAGTT', 'CGACG', 'TCAGA', 'GATAT', 'GAGTT', 'GTGTC', 'AGCTC', 'AAAGC', 'AGTGT', 'CTCGA', 'CGCAC', 'AAGCT', 'CACTG', 'ATCAG', 'GTGTC', 'GTTCC', 'GACGC', 'CTCAC', 'CTCTT', 'GATAT', 'GCGAT', 'CGAAC', 'CGAAC', 'CACTG', 'CGACG', 'CGCAC', 'AGAGT', 'TCACT', 'TCAGA', 'AGTGT', 'GCTCA', 'GATAT', 'TCGAC', 'TCTCG', 'CAGAG', 'GAGTT', 'CACTG', 'TATCA', 'AAAGC', 'CGCAC', 'CGATA', 'GCACT', 'GACGC', 'GACGC', 'AGCGA', 'TCGAC', 'ACGCA', 'CGAAC', 'ATCTC', 'AGTTC', 'CGACG', 'CGACG', 'TCGAC', 'CACTG', 'AGAGT', 'TATCA', 'TCACT', 'GCACT', 'TTCCA', 'GAACT', 'TATCA', 'GATAT', 'AAGCT', 'TCTCG', 'CCAGT', 'ATCAG', 'TCGAC', 'TTCCA', 'AACTC', 'GTTCC', 'TCGAA', 'AAAGC', 'CGCAC', 'TCTCG', 'GCACT', 'AGTTC', 'GAGTT', 'GACGC', 'TCCAG', 'ATCTC', 'GATAT', 'GTTCG', 'CGCAC', 'AGCGA', 'ACGCA', 'CGAAC', 'GCGAT', 'GCACT', 'GTTCG', 'TCGAA', 'TCTCG', 'TCAGA', 'GCACT', 'TCGAC', 'GCACT', 'GCACT', 'AGTGT', 'ACGCA', 'GCACT', 'TCTCG', 'GAACT', 'AGAGT', 'CTCGA', 'TCGAC', 'AGAGT', 'ATCTC', 'AGTTC', 'TCGAA', 'CAGTG', 'ACTCT', 'ATCTC', 'CGCAC', 'TGCGG', 'AACTC', 'CGAAC', 'AGTTC', 'ATCTC', 'TTCGA', 'ACTGC', 'ACTCT', 'TCACT', 'CACTG', 'CGATA', 'TTCGA', 'CACTG', 'TCACT', 'GTTCC', 'TCACT', 'AGCGA', 'GTTCC', 'ACTGC', 'GTTCG', 'CTCTT', 'GAGTT', 'AGAGT', 'CGACG', 'GACGC', 'CTCAC', 'AGAGT', 'GAGTT', 'AACTC', 'TCTCG', 'CTCGA', 'TCAGA', 'CGACG', 'CTCGA', 'GCACT', 'AGTTC', 'ACGCA', 'CGCAC', 'AGCGA', 'AGTGT', 'CTCTT', 'CACTG', 'AGTTC', 'CGACG', 'TCACT', 'TCAGA', 'AGTTC', 'TCGAA', 'GCGAT', 'CGCAC', 'AGTTC', 'GCACT', 'GTGTC', 'TCCAG', 'ACGCA', 'ACGCA', 'GCGAT', 'ACGCA', 'ATCTC', 'AGAGT', 'GCACT', 'GCACT', 'GCGAT', 'GACGC', 'TTCCA', 'TCGAC', 'ATCAG', 'CTCTT', 'ATCTC', 'GCACT', 'TGCGG', 'GATAT', 'GTGTC', 'GTGTC', 'AACTC', 'CGACG', 'TCGAA', 'ATCTC', 'CTCGA', 'CCAGT', 'TCGAC', 'CTCTT', 'GTGTC', 'GATAT', 'GCTCA', 'CGACG', 'GTTCC', 'ACGCA', 'AAGCT', 'CGACG', 'AGAGT', 'TATCA', 'TCGAA', 'TCTCG', 'CCAGT', 'CGACG', 'GAGTT', 'GTTCC', 'TCGAC', 'TCTCG', 'TCGAA', 'GAACT', 'TCGAC', 'CGCAC', 'TCGAC', 'TCTCG', 'AGTGT', 'CACTG', 'GCACT', 'TCGAC', 'AGTTC', 'TATCA', 'ACGCA', 'AGTGT', 'GCACT', 'CGATA', 'CGACG', 'GAGTT', 'CCAGT', 'TCTCG', 'CCAGT', 'CGAAC', 'GTGTC', 'GCACT', 'CTCTT', 'CGATA', 'CTCGA', 'GCGAT', 'CGCAC', 'GCGAT', 'TCTCG', 'TTCGA', 'TCTCG', 'GACGC', 'TCTCG', 'TCGAC', 'ACGCA', 'ACGCA', 'ACGCA', 'CAGAG', 'CACTG', 'GCGAT', 'CACTG', 'TTCGA', 'CACTG', 'ACTCT', 'TTCCA', 'CCAGT', 'GAGTT', 'ACTGC', 'CGCAC', 'GAGTT', 'GACGC', 'CTCTT', 'AGTTC', 'CGCAC', 'TCGAA', 'ACTCT', 'AGAGT', 'GTGTC', 'TCGAC', 'AAGCT', 'CGACG', 'CGATA', 'CGCAC', 'GAGTT', 'TCACT', 'AGTTC', 'ACGCA', 'CGACG', 'ATCTC', 'GCTCA', 'CTGCG', 'CTGCG', 'GAGTT', 'GCGAT', 'TCTCG', 'CGCAC', 'CGACG', 'CCAGT', 'CGACG', 'ACGCA', 'CTCTT', 'CTCGA', 'GAGTT', 'TTCGA', 'AAAGC', 'AAGCT', 'ATCTC', 'TTCCA', 'TTCCA', 'ATCAG', 'GTTCC', 'CTCTT', 'GTTCG', 'GAGTT', 'TCAGA', 'CGCAC', 'AACTC', 'GATAT', 'TGCGG', 'GCGAT', 'ACGCA', 'GTTCG', 'GCTCA', 'AGCGA', 'GTGTC', 'CACTG', 'GCACT', 'GACGC', 'GAGTT', 'CACTG', 'TCACT', 'TCGAC', 'AGTGT', 'GAGTT', 'GCACT', 'CTCGA', 'GCACT', 'GATAT', 'ATCAG', 'ATATC', 'TCGAA', 'CACTG', 'CAGAG', 'AGCTC', 'AGAGT', 'TCACT', 'CGAAC', 'AAGCT', 'AGTGT', 'ACTGC', 'CAGAG', 'GACGC', 'CGCAC', 'GAACT', 'AAAGC', 'GCACT', 'CTCGA', 'CTCTT', 'ATATC', 'ACTGC', 'CTCGA', 'TGCGG', 'CTCAC', 'CTCAC', 'TCGAC', 'GAGTT', 'GAGTT', 'GCACT', 'CAGTG', 'GCACT', 'TCGAC', 'CGATA', 'GAGTT', 'AGCTC', 'CACTG', 'GAGTT', 'GCACT', 'CGACG', 'GATAT', 'GCGAT', 'CACTG', 'CACTG', 'GAACT', 'CACTG', 'TCGAC', 'TCTCG', 'CTCGA', 'CGACG', 'ATATC', 'AGTTC', 'GAACT', 'CTCGA', 'GAACT', 'CGACG', 'TCTCG', 'TCCAG', 'CGACG', 'CAGTG', 'GCTCA', 'GCACT', 'TGCGG', 'GAGTT', 'AGAGT', 'GAACT', 'CGACG', 'GCGAT', 'TCTCG', 'GAGTT', 'CGACG', 'CACTG', 'AGCGA', 'CGACG', 'CACTG', 'CGCAC', 'AGTGT', 'GAGTT', 'ATCTC', 'CGAAC', 'CAGTG', 'AAAGC', 'CTCGA', 'CGACG', 'TCGAA', 'TCGAC', 'ACTGC', 'GCACT', 'ATCAG', 'CTCTT', 'CCAGT', 'CGAAC', 'CGACG', 'TCGAA', 'CGATA', 'CAGTG', 'AGAGT', 'TCGAA', 'TCAGA', 'GACGC', 'GTTCG', 'TGCGG', 'CTGCG', 'TCGAA', 'CAGTG', 'AGAGT', 'CGAAC', 'ACGCA', 'CTCGA', 'CGAAC', 'AGAGT', 'GTGTC', 'GCACT', 'GTTCG', 'CGAAC', 'TCGAC', 'TCGAA', 'AACTC', 'CACTG', 'GTTCC', 'CACTG', 'GACGC', 'GCGAT', 'TCTCG', 'GACGC', 'CGATA', 'ATCAG', 'ACGCA', 'ACGCA', 'ACGCA', 'AGTTC', 'ATCTC', 'CTGCG', 'AGAGT', 'TCTCG', 'CGACG', 'GTTCG', 'ACGCA', 'TCCAG', 'TATCA', 'ATCAG', 'ATCTC', 'TCAGA', 'GCACT', 'CGACG', 'TCGAA', 'ACTCT', 'TGCGG', 'CTCGA', 'GTTCC', 'CTCGA', 'GAACT', 'GATAT', 'TCCAG', 'ACGCA', 'GATAT', 'ACTCT', 'GTTCG', 'CCAGT', 'GTTCG', 'CGCAC', 'GCTCA', 'GATAT', 'CGAAC', 'ACTGC', 'TCCAG', 'AACTC', 'AAAGC', 'TCCAG', 'CGATA', 'AAGCT', 'GACGC', 'GACGC', 'ATCTC', 'TCTCG', 'CAGAG', 'CGACG', 'GAACT', 'CTCAC', 'ACTGC', 'TCTCG', 'CGCAC', 'CGATA', 'GAGTT', 'ACTGC', 'AGTGT', 'TCGAC', 'TGCGG', 'GACGC', 'CTCAC', 'TTCCA', 'TCGAA', 'ACGCA', 'GTTCG', 'ACTCT', 'GCTCA', 'GTGTC', 'GTTCG', 'AACTC', 'CTCTT', 'GCACT', 'AGAGT', 'CGACG', 'CTGCG', 'CGACG', 'ACTGC', 'GTTCG', 'TATCA', 'GAGTT', 'TTCCA', 'CGCAC', 'CTCTT', 'AAAGC', 'GAGTT', 'AGCTC', 'ATCTC', 'GTGTC', 'GTGTC', 'CGATA', 'GCACT', 'CGATA', 'GTGTC', 'ATATC', 'CAGTG', 'AGTTC', 'CGACG', 'GCACT', 'CTCGA', 'CTCAC', 'TCAGA', 'CTCGA', 'GTTCG', 'TTCGA', 'ATCAG', 'CGCAC', 'TCGAC', 'ACTCT', 'GCGAT', 'CGAAC', 'CAGAG', 'CTGCG', 'ATCAG', 'GAGTT', 'CAGAG', 'CGACG', 'ACGCA', 'CGATA', 'CAGAG', 'CTCAC', 'GTTCG', 'CTCGA', 'TTCGA', 'CGAAC', 'AACTC', 'AGAGT', 'TATCA', 'ACTCT', 'GAGTT', 'CGAAC', 'AGTTC', 'GACGC', 'AGAGT', 'ATCTC', 'CACTG', 'GTTCC', 'GACGC', 'CGACG', 'AGTTC', 'GACGC', 'CGCAC', 'CAGTG', 'GATAT', 'ACGCA', 'ACGCA', 'GAGTT', 'ATCTC', 'AAAGC', 'CGACG', 'TCGAA', 'TCGAC', 'AGCGA', 'CGATA', 'AGTTC', 'CTCAC', 'CAGAG', 'AAAGC', 'CAGAG', 'GAGTT', 'CGCAC', 'TATCA', 'GCACT', 'ATCAG', 'AGTTC', 'GCGAT', 'TCGAC', 'AGAGT', 'ATCTC', 'TCAGA', 'TTCCA', 'GTTCG', 'CGACG', 'AGTTC', 'AGAGT', 'GCACT', 'ACGCA', 'CTGCG', 'CTCGA', 'CACTG', 'CACTG', 'CGAAC', 'AGTGT', 'GTGTC', 'CAGTG', 'CACTG', 'TCGAC', 'CGAAC', 'TCTCG', 'GAACT', 'CCAGT', 'CTCAC', 'AGTGT', 'CTCGA', 'GCACT', 'GATAT', 'TTCCA', 'CTCAC', 'GCACT', 'GTTCG', 'TCCAG', 'CCAGT', 'TTCCA', 'CTCGA', 'AGTGT', 'CTCAC', 'TCGAC', 'AGCGA', 'ACGCA', 'GCACT', 'CTCGA', 'ACTGC', 'AGTTC', 'AGTTC', 'AGCTC', 'ATCTC', 'CGCAC', 'CGAAC', 'CGACG', 'GCGAT', 'CTCTT', 'GTTCC', 'CGCAC', 'TTCGA', 'AGAGT', 'CAGTG', 'ATATC', 'GTTCC', 'ATATC', 'CGAAC', 'TCACT', 'CTCTT', 'ACTCT', 'TCGAC', 'GCACT', 'AGTTC', 'TTCCA', 'CGATA', 'AGAGT', 'CTCAC', 'GTTCG', 'CACTG', 'CCAGT', 'TCTCG', 'AAGCT', 'GCACT', 'GAGTT', 'ACTCT', 'TCTCG', 'ACGCA', 'GAACT', 'CTCTT', 'TCGAC', 'AGAGT', 'CTCGA', 'GATAT', 'AGAGT', 'CACTG', 'GAGTT', 'CGCAC', 'CCAGT', 'ACGCA', 'GCACT', 'TTCGA', 'CTCTT', 'CAGAG', 'CACTG', 'GTTCG', 'GAACT', 'GATAT', 'GAGTT', 'CACTG', 'ATCTC', 'ACGCA', 'AGTTC', 'ATCTC', 'GCACT', 'AAGCT', 'GATAT', 'GTGTC', 'TTCGA', 'ACTCT', 'GTGTC', 'CTCGA', 'GAGTT', 'GTTCG', 'GAGTT', 'GTGTC', 'TCACT', 'CGATA', 'AGTTC', 'CGAAC', 'TCCAG', 'CGCAC', 'GTTCG', 'CACTG', 'GTGTC', 'AGCGA', 'AGAGT', 'GAGTT', 'GTGTC', 'TCACT', 'GTTCC', 'TGCGG', 'TCAGA', 'AGCTC', 'GTTCC', 'AGTTC', 'CACTG', 'TCGAA', 'TCCAG', 'TCACT', 'GACGC', 'ATCAG', 'AGTGT', 'TCAGA', 'GTTCC', 'GCACT', 'AGCGA', 'GTGTC', 'AACTC', 'TTCCA', 'CAGTG', 'GTGTC', 'CACTG', 'AGCTC', 'CCAGT', 'GACGC', 'TCACT', 'GCACT', 'GCACT', 'CGACG', 'GTTCG', 'GTTCC', 'CTCTT', 'TCGAA', 'GAGTT', 'GCACT', 'TATCA', 'CTCTT', 'TCGAC', 'CAGTG', 'CTCGA', 'ACTGC', 'GTGTC', 'ACGCA', 'CTCGA', 'ACGCA', 'ATCTC', 'GAACT', 'TCAGA', 'ACGCA', 'CTCTT', 'CGACG', 'TCGAA', 'GCGAT', 'CTGCG', 'ACGCA', 'TCCAG', 'CTCGA', 'CGATA', 'ACGCA', 'GACGC', 'GAGTT', 'ACGCA', 'ACGCA', 'AACTC', 'CTCTT', 'ACGCA', 'CGATA', 'TCTCG', 'AGTGT', 'ATATC', 'AAAGC', 'GTTCG', 'CGACG', 'AGAGT', 'CGCAC', 'CGACG', 'ACTCT', 'TATCA', 'CCAGT', 'CTCTT', 'TCGAA', 'AGCGA', 'CAGAG', 'GCTCA', 'GCACT', 'GCACT', 'TGCGG', 'AGCGA', 'TCTCG', 'AAGCT', 'AAAGC', 'GACGC', 'CTGCG', 'AGAGT', 'TCTCG', 'GTTCC', 'CTCTT', 'AGTTC', 'AGTGT', 'GATAT', 'GTGTC', 'GAGTT', 'TCGAC', 'AGAGT', 'CGACG', 'GAACT', 'AGTTC', 'GACGC', 'CACTG', 'CCAGT', 'TCTCG', 'CGCAC', 'CGACG', 'AGTTC', 'GAACT', 'CTCAC', 'CGACG', 'CTGCG', 'GTTCG', 'TATCA', 'AGTGT', 'TCGAC', 'CCAGT', 'AGCGA', 'ATCTC', 'AGTTC', 'CTCGA', 'AGAGT', 'GCACT', 'AAGCT', 'GCGAT', 'GCTCA', 'CTCTT', 'CAGAG', 'AGAGT', 'CTCAC', 'CGACG', 'GCACT', 'GTTCG', 'GTGTC', 'CTCGA', 'CTCTT', 'AGTTC', 'ACTCT', 'CAGAG', 'AGTTC', 'GACGC', 'GAACT', 'GACGC', 'CGCAC', 'AAGCT', 'AGCTC', 'GCGAT', 'AACTC', 'GAGTT', 'GCACT', 'ACGCA', 'CGCAC', 'GCTCA', 'GCACT', 'GAACT', 'AGAGT', 'GCGAT', 'GCACT', 'CAGTG', 'CGCAC', 'ATCTC', 'GTGTC', 'AGCTC', 'CGCAC', 'TATCA', 'AGCGA', 'CAGAG', 'GCACT', 'CTCGA', 'ACTCT', 'TCAGA']\n"
     ]
    }
   ],
   "source": [
    "N_iter = 100 #number of E/M iterations\n",
    "\n",
    "def find_all_alignments(s, read):\n",
    "    tmp = []\n",
    "    for j in range(len(s)):\n",
    "        if read in s[j]:\n",
    "            tmp.append(j)\n",
    "    return tmp\n",
    "\n",
    "# A: Compatibility Matrix. \n",
    "# Note: You can choose to represent this matrix in whatever form is easiest for your calculations\n",
    "# The form we are pushing you towards here is just how we chose to implement the algorithm\n",
    "A = []\n",
    "for each_read in reads:\n",
    "    A.append( find_all_alignments(transcripts, each_read) )\n",
    "\n",
    "hidden_state_prior = np.zeros([N, M])\n",
    "rho_est = np.array([1/5, 1/5, 1/5, 1/5, 1/5])\n",
    "\n",
    "for i in range(N_iter):\n",
    "    # 1. E step\n",
    "    for i in range(N):\n",
    "        validGenes = A[i]\n",
    "        for j in range(M):\n",
    "            if j in validGenes:\n",
    "                hidden_state_prior[i,j] = rho_est[j]*(1/10)\n",
    "            else:\n",
    "                hidden_state_prior[i,j] = 0    \n",
    "        hidden_state_prior[i] /= sum(hidden_state_prior[i])\n",
    "            \n",
    "\n",
    "        \n",
    "                \n",
    "    # 2. M step\n",
    "    rho_est = hidden_state_prior.sum(axis=0)\n",
    "    rho_est /= sum(rho_est)\n",
    "    \n",
    "print('(est.) rho', rho_est)\n",
    "plt.bar(np.arange(M) + 1, rho_est, color='green', width=0.4, label='Estimated')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Congratulations -- you've just finished your last lab for EE 126!\n",
    "\n",
    "Take a minute to appreciate what you've accomplished throughout the semester. This class wasn't easy, but hopefully you've learned a lot! Come see us if you'd like some help figuring out where to go from here."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "[1] L. Pachter. Models for transcript quantification from RNA-Seq. available at arXiv:1104.3889 [q-bio.GN], 2011. <br />\n",
    "[2] A. Roberts and L. Pachter, RNA-Seq and find: entering the RNA deep field, Genome Medicine, 3 (2011), 74.<br/>\n",
    "[3] C. Do and S. Batzoglou, What is the expectation maximization algorithm? Nature Biotechnology, 26 (2008), 8."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
